R/accessory.R

```diff
@@ -905,70 +905,87 @@ checkListIsSubset <- function(test_list,
 #' @return output Validated contrasts data frame
 #' @export
 
-read_contrasts <-
+read_contrasts <- 
   function(filename,
-           samples,
-           variable_column = "variable",
-           reference_column = "reference",
-           target_column = "target",
-           blocking_column = "blocking",
-           convert_to_list = FALSE) {
-    # Read the contrasts
-
+          samples,
+          variable_column = "variable",
+          reference_column = "reference",
+          target_column = "target",
+          blocking_column = "blocking",
+          convert_to_list = FALSE) {
+
+  # Read the contrasts depending on the file format (CSV or YAML)
+  if (grepl("\\.csv$", filename)) {
     contrasts <- read_metadata(filename)
     contrast_cols <- c(variable_column, reference_column, target_column)
-    if (! blocking_column %in% names(contrasts)){
-        contrasts[[blocking_column]] <- NA
+    if (!blocking_column %in% names(contrasts)) {
+      contrasts[[blocking_column]] <- NA
     }
 
     # Check contrast headers are as expected
-
     if (!all(contrast_cols %in% colnames(contrasts))) {
       stop(paste("Contrasts file must contain all of", paste(contrast_cols, collapse = ", ")))
     }
 
-    # Check contrast content is appropriate to sample sheet
-
-    ## 'variable' values should be sample sheet columns
+  } else if (grepl("\\.ya?ml$", filename)) {
+    contrasts_yaml <- yaml::read_yaml(filename)
+
+    # Extract contrasts and structure into a data frame
+    contrasts_list <- contrasts_yaml$contrasts
+    contrasts <- data.frame(
+      id = sapply(contrasts_list, function(x) x$id),
+      variable = sapply(contrasts_list, function(x) x$comparison[1]),
+      reference = sapply(contrasts_list, function(x) x$comparison[2]),
+      target = sapply(contrasts_list, function(x) x$comparison[3]),
+      blocking = sapply(contrasts_list, function(x) ifelse(!is.null(x$blocking_factors), paste(x$blocking_factors, collapse = ";"), NA)),
+      stringsAsFactors = FALSE
+    )
```

Comment:
```suggestion
    # Parse YAML contrasts into a data frame
    contrasts <- do.call(rbind, lapply(contrasts_yaml$contrasts, function(x) {
      data.frame(
        id = x$id,
        variable = x$comparison[1],
        reference = x$comparison[2],
        target = x$comparison[3],
        blocking = ifelse(is.null(x$blocking_factors), NA, paste(x$blocking_factors, collapse = ";")),
        stringsAsFactors = FALSE
      )
    }))
```

Avoid all the sapplys